### ISSUE TITLE
Propagate trace headers

### ISSUE THREAD

**Issue by seankhliao** [pomerium/pomerium#535]
**Is your feature request related to a problem? Please describe.**

Distributed tracing is useful for tracking requests across a series of services. 
Pomerium already implements tracing through Jaeger for internal comms,
but these are not coordinated with other services

**Describe the solution you'd like**

- Forward Auth: Extract Trace ID from headers (maybe a security concern?)
- Proxy: Inject Trace ID headers into outgoing requests

**Describe alternatives you've considered**

wrap Pomerium in a tracing sideca, but lose visibility into internal tracing

**Explain any additional use-cases**

**Additional context**

"all" is not a very useful name in all-in-one mode, may I suggest "pomerium"

**Comment by desimone** [comment]
@seankhliao 

Interesting idea, I can absolutely see the utility. 

> Forward Auth: Extract Trace ID from headers (maybe a security concern?)

Just to make sure I have this right, this would be extracting the TraceID from the fronting proxy and use that ID for the rest of the Pomerium handled portion of the request? I think this could work, so long as the trace id is explicit, and fails gracefully. I'm not super well versed on all the tracing options out there / how and where they are stored.
 
> Proxy: Inject Trace ID headers into outgoing requests

This is easier. Could just be an optional route flag. 

Thoughts? 

> "all" is not a very useful name in all-in-one mode, may I suggest "pomerium"

🤔 Fair point. I'm open to that change but should probably be tracked in a separate issue.

Thoughts?

\cc @travisgroth

**Comment by seankhliao** [comment]
> Just to make sure I have this right, this would be extracting the TraceID from the fronting proxy and use that ID for the rest of the Pomerium handled portion of the request? 

Yes, I believe the standard format is TraceContext in HTTP headers supported by OpenCensus [here](https://godoc.org/go.opencensus.io/plugin/ochttp/propagation/tracecontext), it [looks like](https://github.com/census-instrumentation/opencensus-go/blob/84d38db2b42a4fc1eebad7b32616e8f4bd48bd76/plugin/ochttp/server.go#L113) it falls back to creating a new span context when none is found

> This is easier. Could just be an optional route flag.

> should probably be tracked in a separate issue.

agree to both

**Comment by desimone** [comment]
Some ... interesting context of how apps can fail on traceID coming from pomerium if the app doesn't expect it ....
https://pomerium-io.slack.com/archives/CK6SVMPU0/p1584491179090500?thread_ts=1584484538.056000&cid=CK6SVMPU0

**Comment by seankhliao** [comment]
that looks like a zipkin/b3 header, which is also fairly common and supported [here](https://godoc.org/go.opencensus.io/plugin/ochttp/propagation/b3)

**Comment by desimone** [comment]
Totally. What's interesting is that that the Trace header is literally crashing JIRA!  Maybe because JIRA is trying to inject the TraceID header and is getting Pomerium's in a form it doesn't expect (com.github.kristofa.brave.IdConversion.convertToLong) and panics.

**Comment by seankhliao** [comment]
I guess JIRA isn't try/catching? But they're supposed to use different header names `traceparent/tracestate` for TraceContext and `x-b3-*` for B3 so I could see why they don't think it should fail

**Comment by desimone** [comment]
\cc @travisgroth #752

**Comment by travisgroth** [comment]
> Proxy: Inject Trace ID headers into outgoing requests

This is now implemented as part of the change over to envoy.  It should work with `b3` type headers.

> "all" is not a very useful name in all-in-one mode, may I suggest "pomerium"

Also changed when we switched to envoy.  

> Forward Auth: Extract Trace ID from headers (maybe a security concern?)

@desimone I believe this will "just work" if we refactor forward auth so that envoy makes the call to authorize.

Today, I think we're missing the propagation because we aren't running the forward auth handler with our metrics middleware.  I think that's an easy enough fix to put in regardless of the outcome of any forward auth changes.
